Fixes #37834 - Set Firmware selection correct

sbernhard opened the pull request:
Without this PR, its is currently really broken. Without this PR, it does always! show 'Automatic' - except while editing hosts. For new/edit of Compute Profile or in the Create Host page - always 'Automatic' is shown. A user would never know what is really configured. This is because of [1].

If you create a host later, 'Automatic' is always displayed, regardless of which firmware is selected in the Compute Profile.

This change makes sure, that the selected firmware on the Compute Profile is the one which was saved previously and represents the value from the database while editing a existing Compute Profile.

The trick is to extend the new_vm() method with a "dry-run" parameter. dry-run is normally set to "true" as the new_vm() method is used at the compute_profile and hosts page for showing the form. The firmware modification should only be done in case the vm is later on really saved.

[1] https://github.com/theforeman/foreman/pull/10321/files#diff-c88b347f9af46e109987241498e9db3c776cc76433ffe3ba6f58ae1d8a74b9adR11


<!---

Thank you for contributing to The Foreman, please read the
[following guide](https://www.theforeman.org/contribute.html), in short:

* [Create an issue](https://projects.theforeman.org/projects/foreman/issues)
* Reference the issue via `Fixes #1234` in the commit message
* Prefer present-tense, imperative-style commit messages
* Mark all strings for translation, see [1]
* Suggest prerequisites for testing and testing scenarios following example above.
* Prepend `[WIP]` for work in progress to prevent bots from triggering actions
* Be patient, we will do our best to take a look as soon as we can
* Explain the purpose of the PR, attach screenshots if applicable
* List all prerequisites for testing (e.g. VMware cluster, two smart proxies...)
* Reviewers often use extensive list of items to check, have a look prior submitting [2]
* Be nice and respectful

1: https://projects.theforeman.org/projects/foreman/wiki/Translating#Translating-for-developers
2: https://github.com/theforeman/foreman/blob/develop/developer_docs/pr_review.asciidoc
-->

sbernhard committed:
Fixes #37834 - Set Firmware selection correct

Without this PR, its is currently really broken. Without this PR, it
does always! show 'Automatic' - except while editing hosts. For
new/edit of Compute Profile or in the Create Host page - always
'Automatic' is shown. A user would never know what is really configured.
This is because of [1].

If you create a host later, 'Automatic' is always displayed, regardless
of which firmware is selected in the Compute Profile.

This change makes sure, that the selected firmware on the Compute
Profile is the one which was saved previously and represents the
value from the database while editing a existing Compute Profile.

The `process_firmware_attributes` is modified so that the methods
`resolve_automatic_firmware` and `normalize_firmware_type` are only used
if a host is created on the compute resource.

[1] https://github.com/theforeman/foreman/pull/10321/files#diff-c88b347f9af46e109987241498e9db3c776cc76433ffe3ba6f58ae1d8a74b9adR11

sbernhard commented:
Another try, similar to https://github.com/theforeman/foreman/pull/10529 to solve the firmware issue.
This PR tries to solve the root cause of the issue, that the original firmware form value is changed while showing / validating the form and not only just before generating the vm.

@nofaralfasi, please have a look. We can work together on a solution and jump on a call if you have further ideas.

nofaralfasi reviewed (commented):
The code looks good overall. It’s not the ideal solution, but it fixes the problem we’re trying to solve.
One small suggestion: I think the `dry_run` variable name could be more descriptive. Maybe something like `new_host`, or another name that better reflects its purpose.
Also, it might help future readers if you could add a short comment explaining why this variable is needed in the first place.
Also, could you update the tests and consider adding a few more for better coverage?

sbernhard commented:
@nofaralfasi changed the name of the flag and changed it from additional function parameter to a attribute. This solved a lot of test issues. Additionally, I have added a comment.

I have adapted the tests. Foreman tests are green now. 

- I tested host create and edit manually with the UI and it worked.
- I tested compute profile create and edit manually with the UI and it worked.
- Tested host creation with existing compute profiles (Automatic, BIOS, UEFI, UEFI SECURE) and it worked. The host uses the correct profile afterwards: 
  * it choose BIOS if Automatic was used
  * it choose UEFI Secure Boot if boot loader was Grub2 UEFI SecureBoot)

sbernhard commented:
Looks like the test issues in Katello are not related. E.g. https://github.com/theforeman/foreman/actions/runs/16202971385/job/45746456820?pr=10600 is failing because of the same issue.

nofaralfasi reviewed (commented):
I feel that adding the `create_vm_ctx` variable in multiple places introduces unnecessary complexity and makes the code harder to follow and maintain. Is there a way we could simplify or centralize it?

sbernhard commented:
> I feel that adding the `create_vm_ctx` variable in multiple places introduces unnecessary complexity and makes the code harder to follow and maintain. Is there a way we could simplify or centralize it?

I don't know how because this flag need to be set so that it will do the magic. Do you have an idea?

sbernhard commented:
all tests have passed.

nofaralfasi commented on app/models/compute_resource.rb:
```diff
@@ -439,9 +439,11 @@ def resolve_automatic_firmware(firmware, firmware_type)
   # @return [Hash] A hash containing the processed firmware attributes.
   def process_firmware_attributes(firmware, firmware_type)
+    attrs = {}
+    attrs[:firmware_unprocessed] = firmware unless firmware.empty?
```
```suggestion
    attrs[:unprocessed_firmware] = firmware if firmware.present?
```

nofaralfasi commented on app/models/compute_resources/foreman/model/libvirt.rb:
```diff
@@ -151,12 +155,16 @@ def new_vm(attr = { })
       opts.merge!(process_firmware_attributes(opts[:firmware], firmware_type))
 
+      firmware_unprocesssed = opts.delete(:firmware_unprocessed).to_s
+      opts[:firmware] = firmware_unprocesssed unless create_vm_ctx || firmware_unprocesssed.empty?
```
firmware_unprocessed is a bit unclear. Consider renaming it to something like unprocessed_firmware or initial_firmware for better readability.
```suggestion
      # initial_firmware is the firmware type assigned either by default or through the compute profile
      initial_firmware = opts.delete(:initial_firmware).to_s
      opts[:firmware] = initial_firmware if !create_vm_ctx && initial_firmware.present?

```

nofaralfasi commented on app/models/compute_resources/foreman/model/libvirt.rb:
```diff
@@ -132,7 +132,11 @@ def template(id)
 
+    # The `create_vm_ctx` flag indicates whether `new_vm` is being called with the intent
+    # to actually create a VM. If the flag is not set, the call may be for purposes like
+    # validation (e.g., triggered by the UI) without provisioning a VM.
```
`create_vm_ctx` is still not clear IMO. How about calling it `creation_mode`?
```suggestion
    # The `creation_mode` flag indicates whether the new_vm method should actually provision a virtual machine.
    # If set to false, the method may be used for validation, preview, or UI-related logic without performing any real provisioning actions.
```

nofaralfasi commented on app/models/compute_resources/foreman/model/libvirt.rb:
```diff
@@ -132,7 +132,11 @@ def template(id)
+    # to actually create a VM. If the flag is not set, the call may be for purposes like
+    # validation (e.g., triggered by the UI) without provisioning a VM.
     def new_vm(attr = { })
+      create_vm_ctx = attr.delete(:create_vm_ctx) || false
```
```suggestion
      creation_mode = attr.delete(:creation_mode) || false
```

nofaralfasi reviewed (commented):
Instead of introducing a new `create_vm_ctx` flag, have you considered using the `host_compute_attrs` method? It only passes a hash during actual VM creation from Foreman, so we might be able to leverage it to simplify the logic.

sbernhard commented:
> Instead of introducing a new `create_vm_ctx` flag, have you considered using the `host_compute_attrs` method? It only passes a hash during actual VM creation from Foreman, so we might be able to leverage it to simplify the logic.

I will try this out tomorrow. 
Other than that, thanks for your suggestions - I have applied them.

sbernhard commented:
> > Instead of introducing a new `create_vm_ctx` flag, have you considered using the `host_compute_attrs` method? It only passes a hash during actual VM creation from Foreman, so we might be able to leverage it to simplify the logic.
> 
> I will try this out tomorrow. Other than that, thanks for your suggestions - I have applied them.

I had a look at the `host_compute_attrs` method. The only thing what would be possible is, to set the `create_mode`attribute in this method - but, actually to set this attribute "just" before the call to`new_vm` is better in my opionion.

nofaralfasi commented:
> I had a look at the `host_compute_attrs` method. The only thing what would be possible is, to set the `create_mode`attribute in this method - but, actually to set this attribute "just" before the call to`new_vm` is better in my opionion.

What about using the existing `:firmware_type` field from `host_compute_attrs` to detect VM creation?

sbernhard commented:
> > I had a look at the `host_compute_attrs` method. The only thing what would be possible is, to set the `create_mode`attribute in this method - but, actually to set this attribute "just" before the call to`new_vm` is better in my opionion.
> 
> What about using the existing `:firmware_type` field from `host_compute_attrs` to detect VM creation?

Isn't this "more hackish" than have a well-documented flag with exactly this use-case?

sbernhard commented:
All checks have passed again. As said, if it works and there are not other, even better ideas to solve the really bad issue, I would vote to fix it for now. Maybey future-foreman-dev has some even better ideas to improve but for now we should use what we have.

nofaralfasi commented:
> Isn't this "more hackish" than have a well-documented flag with exactly this use-case?

That's a fair question, perhaps it’s worth getting a third opinion on this.
@ekohl can we get your take on this?

Problem: Firmware shows “Automatic” instead of the saved value because `new_vm()` processes firmware during form display.
Option 1: Rely on existing fields - `host_compute_attrs` includes `:firmware_type`, which isn’t present during form display, so we detect creation via `attr.key?(:firmware_type)`.
Pros: no new params, leverages existing structure
Cons: implicit detection

Option 2: Add a `create_mode` flag before calling `new_vm`.
Pros: explicit control flow, clear intent
Cons: more code changes

sbernhard commented:
> > Isn't this "more hackish" than have a well-documented flag with exactly this use-case?
> 
> That's a fair question, perhaps it’s worth getting a third opinion on this. @ekohl can we get your take on this?
> 

I had a look at this again. If we would change it to `firmware_type` it would be possible to do all the magic at this step: https://github.com/theforeman/foreman/blob/658ff7ebd6c755b47a5501abb6bf95b74723599b/app/models/compute_resource.rb#L447

This would then depend on the `host_compute_attrs`  -> `firmware_type` and hopefully, this flag will never be changed / adapted. To use the "create_mode" flag is more is clearer and more obvious in my opinion. 

Suggestion: to get this solved, can we discuss this on a video call?

Update: I will try out this solution now.

ekohl commented on app/models/compute_resource.rb:
```diff
@@ -439,10 +439,26 @@ def resolve_automatic_firmware(firmware, firmware_type)
 
-    attrs = generate_secure_boot_settings(firmware)
-    attrs[:firmware] = normalize_firmware_type(firmware)
+    attrs.merge!(generate_secure_boot_settings(firmware))
```
If `attrs` was empty before then why don't you use
```suggestion
    attrs = generate_secure_boot_settings(firmware)
```

Or if you you want a copy because you're modifying it later:

```suggestion
    attrs = generate_secure_boot_settings(firmware).dup
```

But I don't think that' needed. In other words: why did you modify old line 444?

sbernhard replied:
Thanks @ekohl
absolutely correct. It was a left ofter from previous fix. 
I have adapted the code.

ekohl reviewed (commented):
I'm trying to wrap my head around this change, but it's hard. I can't find where the attributes on the compute profile are processed when the compute profile is saved. You describe issues when creating a compute profile, but I don't see any test for that. Perhaps that's the first step?

ekohl commented on app/models/compute_resource.rb:
```diff
@@ -439,10 +439,25 @@ def resolve_automatic_firmware(firmware, firmware_type)
   # @param firmware_type [String] The firmware type based on the provided PXE Loader.
   # @return [Hash] A hash containing the processed firmware attributes.
   def process_firmware_attributes(firmware, firmware_type)
```
I have a feeling what you're in this method doing is wrong. The whole design is:

* Determine the firmware (`resolve_automatic_firmware`)
* Generate compute resource attributes for that firmware (`generate_secure_boot_settings`)
* Set the firmware (`normalize_firmware_type`)

Now you're changing the `firmware` to really be something else after generating the secure boot settings. Feels like that's wrong.

sbernhard replied:
What the PR does now change is, that your described steps are only done if the data is really brought to the compute resource in order to create a host.

ekohl commented on app/models/compute_resource.rb:
```diff
@@ -439,10 +439,25 @@ def resolve_automatic_firmware(firmware, firmware_type)
+      attrs[:firmware] = normalize_firmware_type(firmware)
+    else
+      attrs[:firmware] = initial_firmware
+    end
```
When I read this it feels like it's reinventing `resolve_automatic_firmware`. That method is hard for me to parse, so I prefer this longer version:
```ruby
def resolve_automatic_firmware(firmware, firmware_type)
  if firmware == 'automatic'
    firmware_type.presence || 'bios'
  else
    firmware
  end
end
```

Then we also have `normalize_firmware_type`:
https://github.com/theforeman/foreman/blob/35f74075d9fbe7bb7e58ae943ef5f11facbb9c32/app/models/compute_resource.rb#L417-L424

I was wondering if we should remove the default to `bios` from `resolve_automatic_firmware` and normalize. So effectively:
```ruby
def resolve_automatic_firmware(firmware, firmware_type)
  if firmware == 'automatic'
    firmware_type.presence 
  else
    firmware
  end
end

def process_firmware_attributes(firmware, firmware_type)
  firmware = normalize_firmware_type(resolve_automatic_firmware(firmware, firmware_type))

  attrs = generate_secure_boot_settings(firmware)
  attrs[:firmware] = firmware
end
```

However, that doesn't work because it swallows the secure boot type.

Still, it feels like your fix should be in `resolve_automatic_firmware` and not in `process_firmware_attributes`.

sbernhard replied:
Both `resolve_automatic_firmware` and `normalize_firmware_type` are changing the firmware which would result in values in the form, which do not reflect the real values stored as compute profile.

Therefore, these values should not be called while using the UI - only in case a host is generated and then the real firmware_type need to be generated.

ekohl reviewed (commented):
The initial description in https://github.com/theforeman/foreman/pull/10550#issue-3092763344 is no longer accurate. Is it fair to say that the new approach doesn't select any option by default so that it will be loaded (in the backend) from the compute resource?

I can't quite figure out what it should be, but I've left my notes inline.

ekohl commented on app/models/compute_resource.rb:
```diff
@@ -423,14 +423,14 @@ def normalize_firmware_type(firmware_type)
   def resolve_automatic_firmware(firmware, firmware_type)
     return firmware unless firmware == 'automatic'
-    firmware_type.presence || 'bios'
+    firmware_type || 'bios'
```
I think `.presence` is still important. Not sure if someone ever posts an empty string, but changing that into `nil` is IMHO a good thing.

ekohl commented on app/models/compute_resource.rb:
```diff
@@ -439,11 +439,23 @@ def resolve_automatic_firmware(firmware, firmware_type)
+      firmware = resolve_automatic_firmware(firmware, firmware_type)
+      attrs[:firmware] = normalize_firmware_type(firmware)
+    else
+      attrs[:firmware] = firmware || 'bios'
```
I think we still must normalize the firmware type.

What I'd like to see is tests for the API creation of hosts with the various types. That makes it really clear which attributes you need to send in to create various hosts.

For example, with the following `POST /api/hosts` (https://apidocs.theforeman.org/katello/latest/apidoc/v2/hosts/create.html):
```json
{
  "host": {
    "name": "example",
    "compute_resource_id": 1,
    "pxe_loader": "Grub2 UEFI"
  }
}
```

Assuming compute resource 1 is libvirt or VMware I'd expect it to end up with a UEFI firmware. Similarly, if you select `Grub2 UEFI HTTPS SecureBoot` for the `pxe_loader` it should enable secure boot. Then `PXELinux BIOS` should be BIOS.

I fear your removal of the `automatic` default could break this, but if we had tests in the first place then the answer would be obvious.

Another thing that we have today:

```json
{
  "host": {
    "name": "example",
    "compute_resource_id": 1,
    "pxe_loader": "Grub2 UEFI",
    "compute_attributes": {
      "firmware": "bios"
    }
  }
}
```

Today this throws an error that it's an invalid combination. Does it still after the change?

What about:
```json
{
  "host": {
    "name": "example",
    "compute_resource_id": 1,
    "compute_attributes": {
      "firmware": "something-invalid"
    }
  }
}
```

I think this will be passed to the compute resource. Today this is normalized to `bios` (whether that's right or wrong).

In the category "better late than never", can you create some tests?

sbernhard replied:
I had a look at your suggestion @ekohl and updated my PR. From my perspective, its pretty clear now. 
CC @nofaralfasi

sbernhard replied:
> I think `.presence` is still important. Not sure if someone ever posts an empty string, but changing that into `nil` is IMHO a good thing.

Updated the PR already and added again.

> I think we still must normalize the firmware type.

Actually, I don't think so. 

The normalization is only necessary in case the data is send to the compute resource API. 
The firmware need to stay at the values which are acceptable by the UI - without normalization. 

I have already added some tests: https://github.com/ATIX-AG/foreman/blob/287832d2c102b301eba623194059142b9989ff41/test/models/compute_resource_test.rb

ekohl replied:
> I have already added some tests: https://github.com/ATIX-AG/foreman/blob/287832d2c102b301eba623194059142b9989ff41/test/models/compute_resource_test.rb

These aren't integration tests, which is what I was thinking about.

sbernhard replied:
I tried to add some tests to the hosts_controller. Unfortunately, firmware_type is always empty because setCompute in the orchestration does not run. Maybe you have an idea?

sbernhard replied:
@ekohl I tried multipe ways to run the hosts_controller integration test with libvirt but because of mocks and disabled orchestration and later on due to limited access to the final set firmware/secure_boot setting, this does not make sense. Actually, the unit test of process_firmware_attributes does show pretty good what is going on and that the change works as implemented.

nofaralfasi commented on app/models/compute_resource.rb:
```diff
@@ -439,11 +439,23 @@ def resolve_automatic_firmware(firmware, firmware_type)
+      firmware = resolve_automatic_firmware(firmware, firmware_type)
+      attrs[:firmware] = normalize_firmware_type(firmware)
+    else
+      attrs[:firmware] = firmware || 'bios'
```
Why use `bios` instead of `automatic`? I think it would be better to set the default to `automatic`.

sbernhard replied:
to be "in sync" with 

- https://github.com/ATIX-AG/foreman/blob/develop/app/models/compute_resource.rb#L422
- https://github.com/ATIX-AG/foreman/blob/develop/app/models/compute_resource.rb#L433

nofaralfasi replied:
After processing the user’s input, the default type should be BIOS. For a new form, the default should be Automatic. References:

VMware: https://github.com/nofaralfasi/foreman/blob/develop/app/models/compute_resources/foreman/model/vmware.rb#L823
https://github.com/nofaralfasi/foreman/blob/develop/app/views/compute_resources_vms/form/vmware/_base.html.erb#L10

Libvirt:
https://github.com/nofaralfasi/foreman/blob/develop/app/models/compute_resources/foreman/model/libvirt.rb#L300
https://github.com/nofaralfasi/foreman/blob/develop/app/views/compute_resources_vms/form/libvirt/_base.html.erb#L11

nofaralfasi replied:
Hi @sbernhard, any updates on this?

sbernhard replied:
> Hi @sbernhard, any updates on this?

Sorry, have a look at the last changes. There are 2 commits: 1 with the "original" implementation and the second commit for changing the default from "bios" to "automatic" and the required *-test changes. 

Let me know what you think about. If ok, I can squash it.

nofaralfasi replied:
Hi @sbernhard, yes please squash it and I will merge this PR. Thanks!

nofaralfasi reviewed (approved):
Thank you @sbernhard!!

nlaphine commented:
I know this is "Closed and Merged", but why "After processing the user’s input, the default type should be BIOS"
We live in modern times. EFI is available on every new machine. BIOS is the odd one out now.

Just a question to prompt a change of thinking for the future if I may be so bold.

Thanks
Peter

nofaralfasi commented:
> I know this is "Closed and Merged", but why "After processing the user’s input, the default type should be BIOS" We live in modern times. EFI is available on every new machine. BIOS is the odd one out now.

Hi Peter,
The default is set to BIOS only when no PXE Loader is selected and the Firmware type is set to Automatic. This has been the existing behavior until now, and this PR does not change it.

If we want to change that behavior, we should open a new issue explaining the rationale, discuss it as a team, and then address it in a separate PR.

This PR focuses on fixing a different problem, so we shouldn’t mix the two concerns.
